### ISSUE TITLE
InvalidTemplateParam false positives on `Collection::groupBy()`, `map()`, `mapInto()`, `combine()`

### ISSUE THREAD

**Issue by alies-dev** [psalm/laravel-psalm-plugin#438]
## Description

Psalm reports `InvalidTemplateParam` on common `Collection` method chains (`groupBy()`, `map()`, `mapInto()`, `combine()`, `filter()`, `sort()`). These are false positives — the code is correct, but Psalm's template parameter tracking loses precision through method chains that transform key/value types.

## Environment

- **Psalm**: 6.3.0 (also reproducible on 7.x-beta)
- **psalm-plugin-laravel**: 4.x-dev
- **Laravel**: 12.x
- **PHP**: 8.3+

### 1. `groupBy()` — key type transformation

```php
use Illuminate\Database\Eloquent\Collection;

// $courses is Collection<int, Course>
$courses = Course::query()->get();

// Psalm reports InvalidTemplateParam here
$grouped = $courses->groupBy(
    static fn(Course $course): string => $course->level
);
```

**Expected:** `Collection<string, Collection<int, Course>>` — no error.
**Actual:** `InvalidTemplateParam` — Psalm can't reconcile the key type change from `int` to `string` when the return type uses `static<TGroupKey, static<TKey, TValue>>`.

### 2. `map()` — value type transformation

```php
use Illuminate\Support\Collection;

// $localGroups is Collection<int, LocalGroup>
$mapped = $localGroups->map(
    static fn(LocalGroup $lg) => [
        'localGroup' => $lg,
        'meetups' => $lg->meetups,
    ]
);
```

**Expected:** `Collection<int, array{localGroup: LocalGroup, meetups: ...}>` — no error.
**Actual:** `InvalidTemplateParam` on the `static<TKey, TMapValue>` return.

### 3. `mapInto()` — class transformation

```php
// $results is \Illuminate\Database\Eloquent\Collection<int, LocalGroup>
$mapData = $query->build()->get()->mapInto(MapLocalGroupData::class);
```

**Expected:** `Collection<int, MapLocalGroupData>` — no error.
**Actual:** `InvalidTemplateParam` — `MapLocalGroupData` doesn't satisfy the original collection's `TValue` template constraint.

### 4. `combine()` — using values as keys

```php
// $classes is Collection<int, string>
$knownNotificationClasses = NotificationLog::query()
    ->select('notification_class')
    ->distinct()
    ->get()
    ->pluck('notification_class');

// Psalm reports InvalidTemplateParam
return $knownNotificationClasses->combine($knownNotificationClasses)->all();
```

**Expected:** `Collection<string, string>` — no error.
**Actual:** `InvalidTemplateParam` — `combine()` returns `static<TValue, TCombineValue>` which changes the key type from `int` to `string`, and Psalm flags this.

### 5. Chained operations

```php
$courses->groupBy(static fn(Course $course): string => $course->level)
    ->map(static fn(Collection $coursesInLevel): Collection => $coursesInLevel->sortBy(
        static fn(Course $course): bool => $course->isClosedForEnrollment()
    ));
```

Each step in the chain compounds the `InvalidTemplateParam` — the `groupBy` triggers one, then `map` on the result triggers another.

## Root cause analysis

The issue is that these Collection methods return `static<TNewKey, TNewValue>` where the template params intentionally differ from the original `static<TKey, TValue>`. When Psalm resolves `static` in the context of the original collection, it checks the new template params against the class's declared bounds (`TKey of array-key, TValue`), but the *specific* instantiation's params don't carry through correctly.

Laravel's docblocks are correct — `groupBy` *should* return `static<TGroupKey, static<TKey, TValue>>`. The problem is in how Psalm (or the plugin) validates template parameter consistency through `static` return types on method chains.

The plugin's `stubs/common/Support/Collection.stubphp` doesn't override these methods, so Psalm uses Laravel's own docblocks directly. Adding stubs with more Psalm-friendly type declarations for these methods would likely resolve the issue.

## Suggested fix

Add stubs for `groupBy()`, `map()`, `mapInto()`, `combine()`, `filter()`, and `sort()` in the plugin's `Collection.stubphp` with return types that Psalm can track through chains — potentially using `@return Collection<...>` (concrete) instead of `@return static<...>` where the template params change.

**Comment by alies-dev** [comment]
Verified this is no longer reproducible.

The `InvalidTemplateParam` false positives are gone:

- `groupBy()` and `combine()` — fixed by stubs added in 8ae06fe with conditional return types that correctly handle key type transformations
- `map()` and `mapInto()` — no issue with current Psalm 7; the return types preserve `TKey` so no template bound violation occurs

Tested all four scenarios from the issue description — zero `InvalidTemplateParam` errors.
